build_tools/cmake/easy.sh

```diff
@@ -0,0 +1,291 @@
+  echo "  src <path>   Specify IREE source path. Default: inferred from script path."
+  echo "  ndk <path>   Specify Android NDK path. Configure an Android build."
+  echo "  android_abi <abi> Specify Android ABI. Optional. Default: ${android_abi}." 
+  echo "  android_platform <platform> Specify Android platform version. Optional. Default: ${android_platform}."
+  echo "  docs         Build docs."
+  echo "  tracy        Build with Tracy profiler instrumentation."
+  echo "  py           Build Python bindings."
+  echo "  java         Build Java bindings."
+  echo "  tf           Build TensorFlow compiler frontend."
+  echo "  tflite       Build TFLite compiler frontend."
+  echo "  xla          Build XLA compiler frontend."
+  echo
+  echo "Examples:"
+  echo
+  echo "  Configure and build with ASan, CCache, and Tracy. Note how 'build' comes last!"
+  echo "    $(basename "$0") asan ccache tracy build"
+  echo
+  echo "  Configure and build and Android NDK build, passing the NDK path. Also with CCache and Tracy."
+  echo "    $(basename "$0") ndk ~/android-ndk-r21d ccache tracy build"
+  echo
+  echo "  Configure a Debug build, build only iree_tools_iree-translate:"
+  echo "    $(basename "$0") Debug build iree_tools_iree-translate"
+  echo
+  echo "  Configure, but do not build, a Release build with Python bindings and TensorFlow compiler:"
+  echo "    $(basename "$0") Release py tf"
+  echo
+  echo "  Do not configure, just build iree_tools_iree-translate"
+  echo "    $(basename "$0") build iree_tools_iree-translate"
+
+  exit 1
+fi
+
+# Parse command-line args
+args=("$@")
+for (( i=0; i < $#; i++ ))
+do
+  case "${args[i]}" in
+    noclear) arg_noclear=1;;
+    printonly) arg_printonly=1;;
+    ccache) arg_ccache=1;;
+    asan) arg_asan=1;;
+    msan) arg_msan=1;;
+    tsan) arg_tsan=1;;
+    py) arg_py=1;;
+    java) arg_java=1;;
+    tf) arg_tf=1;;
+    tflite) arg_tflite=1;;
+    xla) arg_xla=1;;
+    tracy) arg_tracy=1;;
+    docs) arg_docs=1;;
+    Debug|Release|RelWithDebInfo|MinSizeRel) arg_build_type="${args[i]}";;
+    ndk) arg_ndk="$(realpath -s "${args[$((i+1))]}")"; i=$((i+1));;
+    android_abi) android_abi="${args[$((i+1))]}"; i=$((i+1));;
+    android_platform) android_platform="${args[$((i+1))]}"; i=$((i+1));;
+    src) arg_src="$(realpath -s "${args[$((i+1))]}")"; i=$((i+1));;
+    build)
+      arg_build=1
+      if [[ $i < $(($# - 1)) ]]
+      then
+        arg_targets="${args[@]:$((i+1))}"
+      fi  
+      break;;
+    *) echo "Error: unkown argument ${args[i]}" >&2; exit 1;;
+  esac
+done
+
+if [[ "${arg_printonly}" != 1 && -f "CMakeLists.txt" ]]
+then
+  echo "Error: We seem to be in a source directory. Please cd into a build directory." >&2
+  exit 1
+fi
+
+if [[ -z "${arg_build_type}" ]]
+then
+  # Sane default build type
+  # TODO: upstream this sane default to IREE's CMakeLists
+  arg_build_type=RelWithDebInfo
+fi
+
+function walk_up_to_iree_root() {
+  path="$(realpath -s "$1")"
```

Comment:
This assumes that someone has iree in a directory called `iree` and that this isn't in a subdirectory called `iree`? Those seem like not great assumptions, especially since iree does have an `iree` subdirectory. Also shouldn't this be `$0`? If we assume it's a git repository you can use `git rev-parse --show-toplevel`

Reply:
Please specify what the expected directory layout is somewhere, e.g.

```
iree/           - project root, $CWD
  build_tools/
  iree/
  scripts/
  third_party/
  ...
iree-build/     - build directory
```

Reply:
Looks like this function was very confusing! Good thing that we won't need it at all --- it's only useful in the context of having this script in-tree, which there seems to be moderate appetite for, and even then, Geoffrey's `git rev-parse --show-toplevel` tip (thanks! with `-C $1` flag, it does what this function does), still removes the need for this function.

Just for the record, this function was not making assumptions as you feared (because it just walks up the tree from the script location so it stops at the first `iree/` and it doesn't matter if there any other sub/super directories also named `iree` around there). It was `$1` because we're dealing with local function parameters not commandline arguments here. And CWD is the build dir, not the source dir.